### PR TITLE
Add a second load more button - potentially more

### PR DESCRIPTION
**Pull request by kul3r4**
We are receiving lots of requests how to implement something similar to infinite scrolling which I know it's not available at the moment, so I am trying to demonstrate what is possible at the moment with amp-bind and amp-list beside the current sample which shows a "show more" just once.

I came up with this approach which uses amp-bind to dinamically change the src name by incrementing a number, but I am not sure how to hide or disable the button when there are no items anymore, can this be an event on `amp-list` where, when there is an error, I could set something on AMP state? @choumx do you have any ideas how to better implement this?

@ericlindley-g FYI
@ithinkihaveacat who added the logic to disable the button when we only had one

### CODE REVIEW

**Comment by aghassemi** [comment]
@kul3r4 Can think of couple of workaround that might work:
1- Button is inside an amp-form that does an xhr to see if there are more items and uses the result of that XHR to decide to hide itself.
2- Every template in `amp-list` contains the button but CSS is used to hide all but the one in the last `li`. Now if backend decided there is no more items left, it can return a `nomore` flag in the json which can also be used in the template to hide the button in the last item.

**Comment by kul3r4** [comment]
@aghassemi Thanks a lot! I'll give it a try with the option 2, I just moved the button inside the template and looks nice, need to work on the backend now.

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

**Comment by sebastianbenz** [comment]
I'm not sure I would recommend this approach. Usually a Load More button requests more items, but in this case it simply returns longer lists, so it doesn't really solve the problem. At least that's my understanding of a load more functionality, if it's purely an UX issue then this approach is fine...

**Comment by kul3r4** [comment]
Agree, this is the only way I know to use `amp-list` to implement a load-more pattern. 
A different approach would be having different calls corresponding different `amp-list` elements and then having the page published server side; in this scenario the page would contain as many `amp-list` as needed.
@ericlindley-g have you thought about this scenario?

**Comment by sebastianbenz** [comment]
I think Ali's first suggestion could be a possible approach.

**Comment by sebastianbenz** [comment]
e.g. if there is no more content available the form could return a 404 which then hides the load more button.

**Comment by kul3r4** [comment]
@sebastianbenz how does this change the fact that the amp-list returns longer lists?

**Comment by sebastianbenz** [comment]
amp-list supports objects as src, you could append the form result to amp-state and render the state in the list (not 100% sure if that works).

**Comment by kul3r4** [comment]
@sebastianbenz Added a second amp-state to hold the items, PTAL

File: src/50_Samples_%26_Templates/Product_Page.html [hunk 1]
```diff
@@ -751,10 +767,22 @@ <h4 class="p2">Products that have been bought together</h4>
...
+          myState: { items: myState.items.concat(event.response.items)},
+          product: {  showMoreButtonDisabled: true }
+        }); submit-success: AMP.setState({
+          myState: { items: myState.items.concat(event.response.items)},
+          product: {timesShowMoreClicked: product.timesShowMoreClicked + 1 }
+        });">
+      <input type="hidden" name="timesShowMoreClicked" value="0" [value]="product.timesShowMoreClicked.toString()">
+      <input type="submit" value="Show more"
+        class="ampstart-btn caps m1 mb3 show"
+        [class] = "(product.showMoreButtonDisabled == true ? 'ampstart-btn caps m1 mb3 hide' : 'ampstart-btn caps m1 mb3 show')">
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Page.html]: Do you need `ampstart-btn caps m1 mb3 ` when the button is hidden?

File: src/json/more_related_products0.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{
+  "items": [{
+    "id": 7,
+    "img": "/img/product1_640x426.jpg",
+    "name": "Apple",
```
Comment (sebastianbenz) [inline src/json/more_related_products0.json]: Can you give the items different names? e.g. Premium Apple

File: src/50_Samples_%26_Templates/Product_Page.html [hunk 3]
```diff
@@ -751,10 +767,22 @@ <h4 class="p2">Products that have been bought together</h4>
         </a>
       </template>
     </amp-list>
-    <!--
-    The button `tap` action sets the value of the `relatedProductsSrc` to an extended list of related products.
-    -->
-    <button [disabled]="listSrcButtonDisabled" class="ampstart-btn caps m1 mb3" id="listSrcButton" on="tap:AMP.setState({listSrcButtonDisabled: true, product: {relatedProductsSrc: '/json/more_related_products.json'}})">Show more</button>
+    <form method="GET"
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Page.html]: Can you document the ampstate and the amp form?

File: src/50_Samples_%26_Templates/Product_Page.html [hunk 4]
```diff
@@ -550,6 +557,8 @@ <h2 class="m1">Apple</h2>
       "selectedSlideForRed": 0,
       "selectedSlideForGolden": 0,
       "selectedSlideForGreen": 0,
+      "timesShowMoreClicked": 0,
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Page.html]: suggestion: rename `timesShowMoreClicked` to `moreItemsPageIndex`

File: src/json/more_related_products0.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (sebastianbenz) [inline src/json/more_related_products0.json]: nit: rename this file to more_related_products_page0 (and the others accordingly)

File: backend/product-browse.go [hunk 6]
```diff
@@ -253,6 +256,40 @@ func handleSearchRequest(w http.ResponseWriter, r *http.Request, page Page) {
 	http.Redirect(w, r, route, http.StatusSeeOther)
 }
 
+func handleLastLoadMoreRequest(w http.ResponseWriter, r *http.Request) {
+	EnableCors(w, r)
+	SetContentTypeJson(w)
+	timesShowMoreClicked := r.URL.Query().Get("timesShowMoreClicked")
+	if timesShowMoreClicked == "1" {
```
Comment (sebastianbenz) [inline backend/product-browse.go]: also rename to moreItemsPageIndex

File: backend/product-browse.go [hunk 7]
```diff
@@ -253,6 +256,40 @@ func handleSearchRequest(w http.ResponseWriter, r *http.Request, page Page) {
...
+		w.WriteHeader(http.StatusNotFound)
+	}
+	productsFile, err := ioutil.ReadFile(buildShowMorePath(timesShowMoreClicked))
+	if err != nil {
+		w.WriteHeader(http.StatusNotFound)
+	}
+	var productsRoot JsonRoot
+	err = json.Unmarshal(productsFile, &productsRoot)
```
Comment (sebastianbenz) [inline backend/product-browse.go]: Why have a marshall/unmarshall step? You could simply write the file contents to the response.

Comment (kul3r4) [reply backend/product-browse.go]: Removed!

File: backend/product-browse.go (hunk 6 above)
Comment (kul3r4) [reply backend/product-browse.go]: done

File: src/json/more_related_products0.json (hunk 2 above)
Comment (kul3r4) [reply src/json/more_related_products0.json]: done

File: src/json/more_related_products0.json (hunk 5 above)
Comment (kul3r4) [reply src/json/more_related_products0.json]: done

File: src/50_Samples_%26_Templates/Product_Page.html (hunk 1 above)
Comment (kul3r4) [reply src/50_Samples_%26_Templates/Product_Page.html]: removed

File: src/50_Samples_%26_Templates/Product_Page.html (hunk 3 above)
Comment (kul3r4) [reply src/50_Samples_%26_Templates/Product_Page.html]: done

File: src/50_Samples_%26_Templates/Product_Page.html (hunk 4 above)
Comment (kul3r4) [reply src/50_Samples_%26_Templates/Product_Page.html]: done

**Comment by kul3r4** [comment]
@sebastianbenz Can we leave the loading indicator for a second iteration?

File: src/50_Samples_%26_Templates/Product_Page.html [hunk 8]
```diff
@@ -752,9 +760,33 @@ <h4 class="p2">Products that have been bought together</h4>
...
+    <form method="GET"
+        action="/json/more_related_products_page"
+        action-xhr="/json/more_related_products_page"
+        target="_top"
+        on="submit-error: AMP.setState({
+          myState: { items: myState.items.concat(event.response.items)},
+          product: {  showMoreButtonDisabled: true }
+        }); submit-success: AMP.setState({
+          myState: { items: myState.items.concat(event.response.items)},
+          product: {moreItemsPageIndex: product.moreItemsPageIndex + 1 }
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Page.html]: just an idea: it is possible to include a boolean var `hasMorePages` in the result? We could use this to hide the show more button.

File: src/50_Samples_%26_Templates/Product_Page.html [hunk 9]
```diff
@@ -752,9 +760,33 @@ <h4 class="p2">Products that have been bought together</h4>
...
+        });">
+      <input type="hidden" name="moreItemsPageIndex" value="0" [value]="product.moreItemsPageIndex.toString()">
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Page.html]: do we need the `.toString()` here?

**Comment by sebastianbenz** [comment]
Np - to add the loading later

File: src/50_Samples_%26_Templates/Product_Page.html (hunk 8 above)
Comment (kul3r4) [reply src/50_Samples_%26_Templates/Product_Page.html]: Added, thanks

File: backend/product-browse.go [hunk 10]
```diff
@@ -83,6 +85,7 @@ func InitProductBrowse() {
 	RegisterSample("samples_templates/product_page", renderProduct)
 	RegisterSampleEndpoint("samples_templates/product_browse_page", SEARCH, handleSearchRequest)
 	http.HandleFunc("/samples_templates/products", handleProductsRequest)
+	http.HandleFunc(SHOW_MORE_PATH, handleLastLoadMoreRequest)
```
Comment (sebastianbenz) [inline backend/product-browse.go]: nit: s/handleLastLoadMoreRequest/handleLoadMoreRequest

File: src/50_Samples_%26_Templates/Product_Page.html [hunk 11]
```diff
@@ -285,11 +286,17 @@
...
       justify-content: flex-start;
     }
+
     .close-gallery-button {
       z-index: 1;
       position: absolute;
       right: 0;
     }
+
+    .listSrcButton {
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Page.html]: do we still need this rule?

File: src/50_Samples_%26_Templates/Product_Page.html [hunk 12]
```diff
@@ -752,9 +760,31 @@ <h4 class="p2">Products that have been bought together</h4>
       </template>
     </amp-list>
     <!--
-    The button `tap` action sets the value of the `relatedProductsSrc` to an extended list of related products.
+    While infinite scrolling is not possible at the moment in AMP, you can use `amp-bind` to dynamically change the
+    src of `amp-list` and add more items to the page.
+    We bind the `src` attribute to the id of an `amp-state` component so that the `amp-list` will use the items
+    from that component as a `src`. We append items to the `amp-state` everytime the user click on the Show more button, see below for more info.
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Page.html]: Can you add a sentence describing the trick you've been using to init the state with the same content as the amp-list.

**Pull request merged**